### PR TITLE
Provide PHP 8.1 support

### PR DESCRIPTION
**Pull request by ihor-sviziev**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

- Declare php 8.1 support in composer.json file
- Ignore platform requirements for php 8.1 for now
- Add ReturnTypeWillChange attribute to fix Deprecated message
- Ignore psalm failure in test
- Allow the only laminas-code ^4.4.0 to fix php 8.1 compatibility (changes from https://github.com/laminas/laminas-code/pull/87 required)

Fixes https://github.com/magento/magento2/issues/34240

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
Not sure how to trigger all tests

**Comment by ihor-sviziev** [comment]
@froschdesign could you explain how to trigger the test run?

**Comment by froschdesign** [comment]
@ihor-sviziev 
Is done automatically. I don't know why it doesn't work.

**Comment by ihor-sviziev** [comment]
ok. let's close and re-open it again, maybe it will work

**Comment by froschdesign** [comment]
@ihor-sviziev 
Works now!

**Comment by froschdesign** [comment]
@ihor-sviziev 
You can drop version 7.3 of PHP, there is no need to [support outdated versions](https://www.php.net/supported-versions).

**Comment by ihor-sviziev** [comment]
@froschdesign looks like for draft PRs tests are not getting ran :(

**Comment by froschdesign** [comment]
@ihor-sviziev 
> looks like for draft PRs tests are not getting ran :(

No, this has nothing to do with the draft status. It's just a little delayed right now and there are some other problems here on GitHub.

**Comment by ihor-sviziev** [comment]
@froschdesign, it looks like laminas-code 4.4.3 still doesn't support PHP 8.1 (but tests are passing with it), but if we're ignoring platform requirements - it allows to install laminas-code 3.5.* that doesn't support even PHP 8.0 and tests are failing.
Should we wait for release with PHP 8.1 in the `laminas-code`?

Another option - increase minimum required version of `laminas-code` to 4.x, but that will drop the PHP 7.3 support - I think that's not expected.

**Comment by froschdesign** [comment]
@ihor-sviziev 
> Another option - increase minimum required version of `laminas-code` to 4.x, but that will drop the PHP 7.3 support - I think that's not expected.

Please see https://github.com/laminas/laminas-server/pull/30#issuecomment-937645687

**Comment by ihor-sviziev** [comment]
ok... now we removed php 7.3 support and phpcs started running on PHP 7.4, and showing deprecations :(  
Updated to "laminas/laminas-coding-standard": "^2.1.4", similar to laminas-code

**Comment by Ocramius** [comment]
Looks like the CS installer didn't really work?

```
 Error: Start tag expected, '<' not found on line 1, column 1

ERROR: Referenced sniff "./vendor/laminas/laminas-coding-standard/ruleset.xml" does not exist
```

**Comment by ihor-sviziev** [comment]
Ok... looks like we should port https://github.com/laminas/laminas-server/pull/7 to 2.10 now.

**Comment by froschdesign** [comment]
@ihor-sviziev 
> Ok... looks like we should port #7 to 2.10 now.

This pull request will be released with the next minor version: 2.11.0

**Comment by ihor-sviziev** [comment]
Just tried to update the coding standard to the latest version and got a lot of failures.
I think that's not a good option.

Is there any way to run the phpcs tests on top of php7.3, while the minimum required PHP version for this package is 7.4?

**Comment by ihor-sviziev** [comment]
Great! I came to the conclusion that the only way to fix current tests - is to remove ignore platform requirements - it will install version laminas-code version 4.4.0 that works fine with php 8.1.
This is currently impossible because the latest version of laminas-code 4.4.3 still does not support PHP 8.1.
_Once stable laminas-code will be released - we can remove the ignore platform requirements and test failure will be fixed._

Another option - increase the minimum required PHP version to PHP 7.4.
I tried to go this way, but it started running phpcs tests with a newer version and started failing. Later, I tried to update the laminas-coding-standards - the PHP warnings and errors got fixed, but the tests themselves started failing as it includes a different set of rules, so we have to port https://github.com/laminas/laminas-server/pull/7. Not sure if it's a good idea to add so many changes in order to add PHP 8.1 support.

**Comment by froschdesign** [comment]
@ihor-sviziev 
The maintainer of this package is @arueckauer, I have added him for a review.

Thanks for the help so far! 👍

**Comment by Ocramius** [comment]
We should pin `laminas-code` to a newer version

File: composer.json
Comment (Ocramius) [inline composer.json]: This won't do - let's bump to `^4.4.0` please

Comment (ihor-sviziev) [reply composer.json]: In this case - the php 7.3 won't be supported and static tests won't be ran 🙁 see explanation above

Comment (Ocramius) [reply composer.json]: That's fine - 7.3 is dead anyway.

Comment (ihor-sviziev) [reply composer.json]: @Ocramius but for doing that we should port https://github.com/laminas/laminas-server/pull/7 to 2.11 branch. Should we do that?

Comment (Ocramius) [reply composer.json]: why is the CS change needed?

Comment (ihor-sviziev) [reply composer.json]: Currently phpcs tests are running on lowest supported version which is currently 7.3. If we increasing to 7.4 - they are starting to fail with php deprecations and errors. If we updating them to version 2 - php deprecations and errors are getting fixed, but the rule set is different, so a lot of new violations appears.
Seems like all them were fixed in #7 in the 3.0.x branch.

So here I see two options:
1. Increase minimum php version to php 7.4 and port changes from #7 to 2.11.x
2. Keep the minimum required php version to 7.3, wait for new version of laminas-code with 8.1 support and remove "ignore platform requirements" flag for php 8.1. In this case, with php 7.3 we will use laminas-cove version ^3.5; with php 7.4-8.0 we'll use laminas-code version ^4.4.0; with php 8.1 we'll use laminas-code the version that starts supporting php 8.1 (not available yet)


One more idea that I had in mind - run `composer require laminas/laminas-code ^4.4.0` AND ignore platform requirements for php 8.1. In this case, we're forcing to not use version ^3.5 with php 8.1.
I just checked what the symfony integration matrix support - it seems like there are no options to run such command for specific php version.

If you have any thoughts - let me know

File: src/Definition.php
Comment (samsonasik) [inline src/Definition.php]: this seems bc break see https://3v4l.org/NmcTG

Comment (ihor-sviziev) [reply src/Definition.php]: Yeah, but without it tests were failing. Any thoughts how can we keep it backward compatible?
PS: these changes already done in the 3.0.x branch, so maybe better to target these changes to it?

Comment (samsonasik) [reply src/Definition.php]: what php version you're using to run tests? I tried run in php 8 and php 7.4 without `int` return type declaration and the tests are green:

Screen Shot 2021-10-08 at 11 09 07

Comment (ihor-sviziev) [reply src/Definition.php]: It's failing on 8.1 :)

Comment (samsonasik) [reply src/Definition.php]: I tried in php `8.1.0RC4-dev` and it shows green:

Screen Shot 2021-10-08 at 11 22 11

are it error on github action with `8.1.0RC3-dev` ?

Comment (ihor-sviziev) [reply src/Definition.php]: Yes. Unfortunately, I did a force push and now it will be not so easy to find. We can temporary try to revert or and see

Comment (ihor-sviziev) [reply src/Definition.php]:
```suggestion
    public function count()
```

**Comment by samsonasik** [comment]
Looking at the error at php 8.1 rc3, It seems you can add: `#[\ReturnTypeWillChange]` on the `count()` method.

**Comment by ihor-sviziev** [comment]
@samsonasik it did the trick! Thank you!
A bit later I'll squash all changes

File: composer.json
Comment (ihor-sviziev) [reply composer.json]: Ok, now I found a way to ignore the PHP 8.1 job with lowest dependencies and committed as https://github.com/laminas/laminas-server/pull/30/commits/4a270b35501d726ad6405397447ba88a2f0cd470. I think it's good enough solution for now

**Review by arueckauer — APPROVED** [review]
Thank you very much, @ihor-sviziev ! 

LGTM

Thanks @froschdesign, @Ocramius and @samsonasik  for reviewing and providing feedback, greatly appreciated!

**Comment by froschdesign** [comment]
@arueckauer 
The branch for 2.11.x is missing: https://github.com/laminas/laminas-server/branches

**Pull request merged**
